File: services/frontend/src/services/backendApi/compile.ts
/**
 * LaTeX 编译与 SyncTeX 相关 API。
 */

import { http, buildHeaders, BackendError, BASE } from './client'

export interface CompilerInfo {
  available: string[]
  default: string
}

export interface CompileRequest {
  compiler?: string
  main_doc_id?: string
  incremental_compile?: boolean
  from_scratch?: boolean
  is_auto_compile?: boolean
}

export type CompileStatus =
  | 'success'
  | 'failure'
  | 'compile-in-progress'
  | 'too-recently-compiled'
  | 'autocompile-backoff'
  | 'unavailable'
  | 'timedout'

export interface CompileResult {
  ok: boolean
  status: CompileStatus
  build_id: string
  compiler: string
  duration_ms: number
  error: string
  log_tail: string
  pdf_bytes: number
}

export interface CompileSyncToPdfRequest {
  document_id: string
  offset: number
}

export interface CompileSyncToPdfResult {
  page: number
  x: number
  y: number
  width: number | null
  height: number | null
  line: number
  column: number
}

export interface CompileSyncFromPdfRequest {
  page: number
  x: number
  y: number
}

export interface CompileSyncFromPdfResult {
  document_id: string
  offset: number
  line: number
  column: number
  source_path: string
}

export interface CompileSettings {
  main_doc_id: string
  compiler: string
  incremental_compile: boolean
}

export const compileApi = {
  listCompilers: () => http<CompilerInfo>('/api/compile/compilers'),
  rescanCompilers: () =>
    http<CompilerInfo>('/api/compile/rescan', { method: 'POST' }),
  compile: (body?: CompileRequest) =>
    http<CompileResult>('/api/compile', {
      method: 'POST',
      body: JSON.stringify(body ?? {}),
    }),
  syncToPdf: (body: CompileSyncToPdfRequest) =>
    http<CompileSyncToPdfResult>('/api/compile/sync-to-pdf', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  syncFromPdf: (body: CompileSyncFromPdfRequest) =>
    http<CompileSyncFromPdfResult>('/api/compile/sync-from-pdf', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  pdfUrl: (projectId: string) => `${BASE}/api/projects/${projectId}/compile.pdf`,
  getLog: async () => {
    const resp = await fetch(`${BASE}/api/compile/log`, {
      credentials: 'include',
      headers: buildHeaders(),
    })
    if (!resp.ok) throw new BackendError(resp.status, resp.statusText)
    return resp.text()
  },
  getSettings: () => http<CompileSettings>('/api/compile/settings'),
  updateSettings: (body: Partial<CompileSettings>) =>
    http<CompileSettings>('/api/compile/settings', {
      method: 'PUT',
      body: JSON.stringify(body),
    }),
  clearCache: () =>
    http<void>('/api/compile/cache', {
      method: 'DELETE',
    }),
}

// Conversations / chat

File: services/backend/app/schemas/compile.py
"""LaTeX 编译与 SyncTeX schema。"""

from __future__ import annotations

from pydantic import BaseModel


class CompilerInfoOut(BaseModel):
    available: list[str]
    default: str


class CompileIn(BaseModel):
    compiler: str | None = None
    main_doc_id: str | None = None


class CompileOut(BaseModel):
    ok: bool
    compiler: str
    duration_ms: int
    error: str
    # Truncated log preview. Full log is fetched via /api/compile/log.
    log_tail: str
    # Length of the PDF blob in bytes (0 if no PDF).
    pdf_bytes: int


class CompileSyncToPdfIn(BaseModel):
    document_id: str
    offset: int


class CompileSyncToPdfOut(BaseModel):
    page: int
    x: float
    y: float
    width: float | None = None
    height: float | None = None
    line: int
    column: int


class CompileSyncFromPdfIn(BaseModel):
    page: int
    x: float
    y: float


class CompileSyncFromPdfOut(BaseModel):
    document_id: str
    offset: int
    line: int
    column: int
    source_path: str
